### ISSUE TITLE
Service:AmazonCloudFormation, Message:Template format error: Parameter 'EnvironmentVariables' default value '[****]' length is greater than 4096.

### ISSUE THREAD

**Issue by theodorDiaconu** [zodern/mup-aws-beanstalk#42]
```
Service:AmazonCloudFormation, Message:Template format error: Parameter 'EnvironmentVariables' default value '[****]' length is greater than 4096.
```

I've checked and tried with only one env: {} variable. Any idea what this is about ?

**Comment by theodorDiaconu** [comment]
I think this happens in this app because the settings file is just to big, and mup transforms it into an env variable.

**Comment by theodorDiaconu** [comment]
Applied for a service limit increase "Limit Increase: Elastic Beanstalk", if it works after that, will close the issue.

**Comment by theodorDiaconu** [comment]
This is the response I got:
```
I have checked the environments that are in Terminated status in ap-northeast-1 region and found the following error message:
"Service:AmazonCloudFormation, Message:Template format error: Parameter 'EnvironmentVariables' default value '[****]' length is greater than 4096."

The reason behind the error is that the total size of all properties cannot exceed 4096 bytes, as mentioned here:  https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/environments-cfg-softwaresettings.html#environments-cfg-softwaresettings-console 

As a workaround, you can explore any of the following 2 options:

1. Leveraging Systems Manager Parameter Store [1][2], and integrate in your code the API call GetParameters [3] to retrieve them;

2. Use the mechanism of creating a S3 bucket that can only be accessed by the application, create a file with the parameters, and retrieve the file to be used by your application after. This section of our documentation [4] explains how to implement it, focused on private keys for SSL configuration, but can be used as a starting point as well to secure your parameters into a file, and configure Elastic Beanstalk to securely retrieve it.

For more information about Customizing Software on Linux Servers on Elastic Beanstalk, please refer to this documentation [5]. 

1: https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-paramstore.html
2: https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-paramstore-working.html
3: https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_GetParameters.html
4:https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/https-storingprivatekeys.html
5: https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/customize-containers-ec2.html

I hope this helps, but let us know if you have any questions. 
```

**Comment by zodern** [comment]
Thanks @theodorDiaconu for sharing the response.

I have started working on a fix using s3 to store the env variables in the [long-env-vars](https://github.com/zodern/mup-aws-beanstalk/tree/long-env-vars) branch. I will finish it next week.

The feature is optional for now, and will be enabled by setting `app.longEnvVars` in the config to `true`. 

- [x] Make sure it can switch to storing the env variables in s3 when the bundle created during the last deploy didn't have the additional file needed for this to work.
- [ ] Download the file. This is mostly done, but should be simplified.
- [ ] Use the env variables when starting the app
- [ ] Rollback the changes if the app fails to run with the new env variables. This is partially implemented.
- [ ] Update the validator and docs
- [ ] Cleanup old env files on s3

**Comment by theodorDiaconu** [comment]
@zodern that's amazing!

**Comment by theodorDiaconu** [comment]
@zodern is there anything we could do to speed-up the process ? I've seen some people just prepend the settings into their build-script, and problem solved. But I don't know how that would work with restarting the servers with a new settings file..

**Comment by zodern** [comment]
This is available in 0.6.0-beta5. You can try the beta with `npm i -g mup-aws-beanstalk@next`.

Enable this by setting `app.longEnvVars` to `true`, and run `mup deploy`.

**Comment by sarathbabugit** [comment]
> Applied for a service limit increase "Limit Increase: Elastic Beanstalk", if it works after that, will close the issue.

where you Applied for a service limit increase?
